// models/product.js
const mongoose = require("mongoose");
const Joi = require("joi");

const productSchema = new mongoose.Schema({
  name: { type: String, required: true, minlength: 3, maxlength: 100 },
  price: { type: Number, required: true, min: 0 },
  category: { type: String, required: true, minlength: 3, maxlength: 50 },
  stock: { type: Number, required: true },
  description: { type: String, minlength: 10, maxlength: 500 },
  image: { type: Buffer },
}, { timestamps: true });

const validateProduct = (data) => {
  const schema = Joi.object({
    name: Joi.string().min(3).max(100).required(),
    price: Joi.number().min(0).required(),
    category: Joi.string().min(3).max(50).required(),
    stock: Joi.number().required(),
    description: Joi.string().min(10).max(500).optional(),
    image: Joi.string().optional(),
  });
  return schema.validate(data);
};

console.log("Registering Product model...");
const Product = mongoose.model("Product", productSchema);
console.log("Product model registered successfully");

module.exports = { Product, validateProduct };
